File: python/1310. XOR Queries of a Subarray.py
class Solution(object):
    def xorQueries(self, arr, queries):
        """
        :type arr: List[int]
        :type queries: List[List[int]]
        :rtype: List[int]
        """
        pre = [0] * (len(arr) + 1)
        for i, e in enumerate(arr):
            pre[i + 1] = pre[i] ^ e
        return [pre[a] ^ pre[b+1] for a, b in queries]

# 由
# a ^ b = b ^ a
# 0 ^ x = x
# x ^ x = 0
# 易知：
# d ^ e ^ f =
#  0   ^    (d ^ e ^ f) = 
# ((a ^ a)    ^    (b ^ b)    ^   (c ^ c))      ^      (d ^ e ^ f) = 
# (a ^ b ^ c)   ^   (a ^ b ^ c ^ d ^ e ^ f)

File: go/src/leetcode/1310. XOR Queries of a Subarray/1310. XOR Queries of a Subarray.go
package main

import "fmt"

func xorQueries(arr []int, queries [][]int) []int {
	ans := make([]int, len(queries))
	pre := make([]int, len(arr)+1)
	for i := 1; i <= len(arr); i++ {
		pre[i] = pre[i-1]^arr[i-1]
	}
	for i, q := range queries {
		l, r := q[0], q[1]
		ans[i] = pre[r+1] ^ pre[l]
	}
	return ans
}

func main() {
	fmt.Println(xorQueries([]int{1,3,4,8}, [][]int{{0,1},{1,2},{0,3},{3,3}}))
	fmt.Println(xorQueries([]int{4,8,2,10}, [][]int{{2,3},{1,3},{0,0},{0,3}}))
}